compiler/hash-pipeline/src/workspace.rs

```diff
@@ -18,6 +19,121 @@ use hash_ast::{
+        /// Indicates the source has not undergone any stages after parsing.
+        const NONE = 0;
+
+        /// If set, the compiler will no perform desugaring on the module.
+        const DESUGARED = 0b00000001;
+
+        /// If set, the compiler will not perform ast expansion on the module.
+        const EXPANDED  = 0b00000010;
+
+        /// If set, the compiler will not perform semantic analysis on the module.
+        const CHECKED_SEMANTICS  = 0b00000100;
+
+        /// If set, the compiler will not perform type checking on the module.
+        const TYPECHECKED = 0b00001000;
+
+        /// If set, the compiler will not perform lowering on the module.
+        const LOWERED   = 0b00010000;
+    }
+}
+
+impl SourceStageInfo {
+    /// Check if the source has undergone AST desugaring.
+    pub fn is_desugared(&self) -> bool {
+        self.contains(SourceStageInfo::DESUGARED)
+    }
+
+    /// Check if the source has undergone AST expansion.
+    pub fn is_expanded(&self) -> bool {
+        self.contains(SourceStageInfo::EXPANDED)
+    }
+
+    /// Check if the source has been type checked.
+    pub fn is_typechecked(&self) -> bool {
+        self.contains(SourceStageInfo::TYPECHECKED)
+    }
+
+    /// Check if the source has gone through semantic analysis.
+    pub fn is_semantics_checked(&self) -> bool {
+        self.contains(SourceStageInfo::CHECKED_SEMANTICS)
+    }
+
+    /// Returns true if the source has been lowered.
+    pub fn is_lowered(&self) -> bool {
+        self.contains(SourceStageInfo::LOWERED)
+    }
+
+    /// Set the desugaring flag.
+    pub fn set_desugared(&mut self) {
+        self.insert(SourceStageInfo::DESUGARED);
+    }
+
+    /// Set the expansion flag.
+    pub fn set_expanded(&mut self) {
+        self.insert(SourceStageInfo::EXPANDED);
+    }
+
+    /// Set the type checking flag.
+    pub fn set_typechecked(&mut self) {
+        self.insert(SourceStageInfo::TYPECHECKED);
+    }
+
+    /// Set the semantic analysis flag.
+    pub fn set_checked_semantics(&mut self) {
+        self.insert(SourceStageInfo::CHECKED_SEMANTICS);
+    }
+
+    /// Set the lowering flag.
+    pub fn set_lowered(&mut self) {
+        self.insert(SourceStageInfo::LOWERED);
+    }
+}
+
+/// A map of [SourceId]s to their corresponding [SourceStageInfo]. This is used
+/// to track the current stage of the compiler pipeline for each source.
+#[derive(Debug)]
+pub struct StageInfo(HashMap<SourceId, SourceStageInfo>);
+
+impl StageInfo {
+    /// Create a new [StageInfo] with no sources.
+    pub fn new() -> Self {
+        Self(HashMap::new())
+    }
+
+    /// Add a new source to the [SourceStageInfo] with the given [SourceId].
+    pub fn add(&mut self, source_id: SourceId, stage: SourceStageInfo) {
+        self.0.insert(source_id, stage);
+    }
+
+    /// Update the [SourceStageInfo] for a particular module.
+    pub fn update(&mut self, source: SourceId, info: fn(SourceStageInfo) -> SourceStageInfo) {
```

Comment:
Would be nice to take a closure rather than fn pointer

Reply:
So the type would be `impl FnOnce(...) -> ...` ?

Reply:
yeah